AmplifyPlugins/API/AWSAPICategoryPlugin/Interceptor/SubscriptionInterceptor/IAMAuthInterceptor.swift

```diff
@@ -28,193 +21,137 @@ class IAMAuthInterceptor: AuthInterceptor {
+            let payloadQuery = URLQueryItem(name: RealtimeProviderConstants.payload, value: payloadBase64)
+            urlComponents.queryItems = [headerQuery, payloadQuery]
+            guard let signedUrl = urlComponents.url else {
+                return request
+            }
+            let signedRequest = AppSyncConnectionRequest(url: signedUrl)
+            return signedRequest
         }
-        var intValue: Int?
-        init?(intValue: Int) {
-            // We are not using this, thus just return nil. If we don't return nil, then it is expected all of the
-            // stored properties are initialized, forcing the implementation to have logic that maintains the two
-            // properties `stringValue` and `intValue`. Since we don't have a string representation of an int value
-            // and aren't using int values for determining the coding key, then simply return nil since the encoder
-            // will always pass in the header key string.
-            self.intValue = intValue
-            self.stringValue = ""
 
-        }
-    }
+    func getAuthHeader(
+        _ endpoint: URL,
+        with payload: String) -> IAMAuthenticationHeader? {
 
-    override func encode(to encoder: Encoder) throws {
-        var container = encoder.container(keyedBy: DynamicCodingKeys.self)
-        // Force unwrapping when creating a `DynamicCodingKeys` will always be successful since the string constructor
-        // will never return nil even though the constructor is optional (conformance to CodingKey).
-        try container.encode(authorization,
-                             forKey: DynamicCodingKeys(stringValue: SubscriptionConstants.authorizationkey)!)
-        try container.encode(securityToken,
-                             forKey: DynamicCodingKeys(stringValue: RealtimeProviderConstants.iamSecurityTokenKey)!)
-        try container.encode(amzDate,
-                             forKey: DynamicCodingKeys(stringValue: RealtimeProviderConstants.amzDate)!)
-        try container.encode(accept,
-                             forKey: DynamicCodingKeys(stringValue: RealtimeProviderConstants.acceptKey)!)
-        try container.encode(contentEncoding,
-                             forKey: DynamicCodingKeys(stringValue: RealtimeProviderConstants.contentEncodingKey)!)
-        try container.encode(contentType,
-                             forKey: DynamicCodingKeys(stringValue: RealtimeProviderConstants.contentTypeKey)!)
-        if let headers = additionalHeaders {
-            for (key, value) in headers {
-                try container.encode(value, forKey: DynamicCodingKeys(stringValue: key)!)
+            guard let helper = HeaderIAMSigningHelper(
+                endpoint: endpoint,
+                payload: payload,
+                region: region) else {
+                return nil
             }
+            var signedHeader: IAMAuthenticationHeader?
+            let semaphore = DispatchSemaphore(value: 0)
```

Comment:
Can this cause threading issues?

Reply:
Yes it does, this logic keeps the old behavior. But there is another method that adds new async flow without the semaphore logic.